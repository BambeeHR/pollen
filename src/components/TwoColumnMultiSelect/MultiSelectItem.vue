<template>
  <label
    class="multi-select-item"
    :class="{ focused: isFocused }"
    @mouseover="isHovering = true"
    @mouseleave="isHovering = false"
  >
    <input
      type="checkbox"
      :value="value"
      :checked="checked"
      class="multi-select-item__input"
      :name="name"
      @blur="isFocused = false"
      @focus="isFocused = true"
      @input="$emit('input', $event)"
    />
    <BaseIcon v-if="icon" :icon="icon" class="multi-select-item__pre-icon" />
    <span class="multi-select-item__label">{{ label }}</span>
    <BaseIcon
      v-if="isHovering || isFocused"
      class="multi-select-item__post-icon"
      :icon="checked ? 'close' : 'plus'"
    />
  </label>
</template>

<script>
import BaseIcon from '../BaseIcon/BaseIcon.vue';

export default {
  components: { BaseIcon },
  props: {
    checked: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: null,
    },
    label: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      isFocused: false,
      isHovering: false,
    };
  },
};
</script>

<style scoped>
.multi-select-item {
  @apply cursor-pointer 
    flex 
    items-center
    pl-6
    pr-12 
    py-4
    relative
    text-sm;
}

.multi-select-item.focused {
  @apply shadow-outline;
}

.multi-select-item__input {
  @apply sr-only;
}

.multi-select-item .base-icon {
  @apply text-2xl;
}

.multi-select-item .multi-select-item__pre-icon {
  @apply mr-4;
}

.multi-select-item >>> .material-design-icon__svg {
  @apply bottom-0;
}

.multi-select-item .multi-select-item__post-icon {
  @apply absolute mr-6 right-0;
}
</style>
