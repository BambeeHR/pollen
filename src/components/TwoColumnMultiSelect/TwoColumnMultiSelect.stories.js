import BaseIcon from '../BaseIcon/BaseIcon.vue';
import TwoColumnMultiSelect from './TwoColumnMultiSelect.vue';
import '../../assets/tailwind.css';

export default {
  title: 'TwoColumnMultiSelect',
  component: TwoColumnMultiSelect,
};

export const Default = () => ({
  components: { TwoColumnMultiSelect },
  data() {
    return {
      options: [
        {
          icon: 'folder',
          label: 'Option 1',
          value: 'Option 1',
        },
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
        {
          label: 'Test',
          value: 'test',
        },
        {
          label: 'Something',
          value: 'Something',
        },
        {
          label: 'Hello',
          value: 'Worl',
        },
      ],
      value: [],
    };
  },
  template: `
    <TwoColumnMultiSelect v-model="value" :options="options" />
  `,
});

export const WithFilter = () => ({
  components: { TwoColumnMultiSelect },
  data() {
    return {
      options: [
        {
          label: 'Option 1',
          value: 'Option 1',
        },
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: [],
    };
  },
  template: `
    <TwoColumnMultiSelect v-model="value" :options="options" filterable />
  `,
});

export const WithPrefilledFilter = () => ({
  components: { TwoColumnMultiSelect },
  data() {
    return {
      options: [
        {
          label: 'Option 1',
          value: 'Option 1',
        },
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: [],
      filter: '1',
    };
  },
  template: `
    <TwoColumnMultiSelect 
      v-model="value" 
      :options="options" 
      :initialFilter="filter" 
      filterable 
    />
  `,
});

export const WithHeaderAsProps = () => ({
  components: { TwoColumnMultiSelect },
  data() {
    return {
      options: [
        {
          icon: 'folderOutline',
          label: 'Option 1',
          value: 'Option 1',
        },
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: [],
    };
  },
  template: `
    <TwoColumnMultiSelect
      v-model="value"
      :options="options"
      selectedHeader="Yup"
      unselectedHeader="Nope"
    />
  `,
});
export const WithHeadersAsSlots = () => ({
  components: { BaseIcon, TwoColumnMultiSelect },
  data() {
    return {
      options: [
        {
          label: 'Option 1',
          value: 'Option 1',
        },
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: [],
    };
  },
  template: `
    <TwoColumnMultiSelect v-model="value" :options="options">
      <template v-slot:unselected-header>
        <span :style="{ 'margin-right': '8px' }"><BaseIcon icon="folder" /></span>
        <b>Things left to do...</b>
      </template>
      <template v-slot:selected-header>
        <span :style="{ 'margin-right': '8px' }"><BaseIcon icon="folder" /></span>
        <b>Things that I have done!</b>
      </template>
    </TwoColumnMultiSelect />
  `,
});
