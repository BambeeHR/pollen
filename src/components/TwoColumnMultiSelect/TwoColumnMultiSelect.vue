<template>
  <div class="two-column-multi-select">
    <div class="two-column-multi-select__unselected-items">
      <div class="two-column-multi-select__column-header">
        <slot name="unselected-header">{{ unselectedHeader }}</slot>
      </div>
      <ul class="two-column-multi-select__item-list">
        <li v-if="filterable">
          <SearchField v-model.trim="fitler" />
        </li>
        <li v-for="item in unselectedOptions" :key="item.value">
          <MultiSelectItem v-bind="item" :name="name" @input="handleInput" />
        </li>
      </ul>
    </div>
    <div class="two-column-multi-select__selected-items">
      <div class="two-column-multi-select__column-header">
        <slot name="selected-header">{{ selectedHeader }}</slot>
      </div>
      <ul class="two-column-multi-select__item-list">
        <li v-for="item in selectedOptions" :key="item.value">
          <MultiSelectItem
            v-bind="item"
            :name="name"
            checked
            @input="handleInput"
          />
        </li>
      </ul>
    </div>
  </div>
</template>

<script>
import shortid from 'shortid';
import MultiSelectItem from './MultiSelectItem.vue';
import SearchField from '../SearchField/SearchField.vue';

export default {
  components: { MultiSelectItem, SearchField },
  props: {
    // Options are an array of objects with a label, value, and icon property.
    // All are strings. Icon is optional.
    options: {
      type: Array,
      default: () => [],
    },
    value: {
      type: Array,
      default: () => [],
    },
    name: {
      type: String,
      default: shortid.generate,
    },
    filterable: {
      type: Boolean,
      default: false,
    },
    selectedHeader: {
      type: String,
      default: 'Selected',
    },
    unselectedHeader: {
      type: String,
      default: 'Options',
    },
  },
  data() {
    return {
      filter: '',
    };
  },
  computed: {
    selectedOptions() {
      // Using `map` rather than `filter` here preserves the order of `value`.
      return this.value.map((value) =>
        this.options.find((item) => item.value === value)
      );
    },
    unselectedOptions() {
      return this.options.filter(
        (item) =>
          !this.value.includes(item.value) &&
          (!this.filterable ||
            (this.filterable &&
              item.label
                .toLowerCase()
                .includes(this.filter.toLowerCase().trim())))
      );
    },
  },
  methods: {
    handleInput(e) {
      const { checked, value } = e.target;
      const newValue = checked
        ? [...this.value, value]
        : this.value.filter((item) => item !== value);
      this.$emit('input', newValue);
    },
  },
};
</script>

<style scoped>
.two-column-multi-select {
  @apply border
    border-gray-5
    flex
    /* For legacy */
    border-solid;
}

.two-column-multi-select__unselected-items,
.two-column-multi-select__selected-items {
  @apply flex flex-col flex-grow w-1/2;
}

.two-column-multi-select__unselected-items {
  @apply border-r
    border-gray-5
    /* For legacy */
    border-b-0
    border-l-0
    border-t-0
    border-solid;
}

.two-column-multi-select__item-list {
  @apply flex-grow
    /* For legacy */
    list-none
    m-0
    p-0;
}

.two-column-multi-select__column-header {
  @apply border-b
    border-gray-5
    font-body-small
    font-semibold
    px-6
    py-3
    /* For legacy */
    border-l-0
    border-r-0
    border-t-0
    border-solid;
}

.two-column-multi-select .search-field {
  @apply rounded-none;
}
</style>
