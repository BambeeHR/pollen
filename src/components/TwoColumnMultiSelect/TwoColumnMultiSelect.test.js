import { shallowMount } from '@vue/test-utils';
import TwoColumnMultiSelect from './TwoColumnMultiSelect.vue';

const options = [
  {
    icon: 'folderOutline',
    label: 'Option 1',
    value: 'Option 1',
  },
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
];

describe('TwoColumnMultiSelect', () => {
  test('renders correctly', () => {
    const wrapper = shallowMount(TwoColumnMultiSelect, {
      propsData: {
        options,
        name: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
