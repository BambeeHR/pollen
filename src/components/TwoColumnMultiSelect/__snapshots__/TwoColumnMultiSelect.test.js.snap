// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoColumnMultiSelect renders correctly 1`] = `
<div
  class="two-column-multi-select"
>
  <div
    class="two-column-multi-select__unselected-items"
  >
    <div
      class="two-column-multi-select__column-header"
    >
      Options
    </div>
     
    <ul
      class="two-column-multi-select__item-list"
    >
      <!---->
       
      <li>
        <multiselectitem-stub
          icon="folderOutline"
          label="Option 1"
          name="test"
          value="Option 1"
        />
      </li>
      <li>
        <multiselectitem-stub
          label="Option 2"
          name="test"
          value="foo"
        />
      </li>
      <li>
        <multiselectitem-stub
          label="Option 3"
          name="test"
          value="bar"
        />
      </li>
    </ul>
  </div>
   
  <div
    class="two-column-multi-select__selected-items"
  >
    <div
      class="two-column-multi-select__column-header"
    >
      Selected
    </div>
     
    <ul
      class="two-column-multi-select__item-list"
    />
  </div>
</div>
`;
